[SPARK-6222][STREAMING] Make sure batches are considered fully processed only after the checkpoint after batch is completed.

harishreedharan opened the pull request:
Streaming jobs start a checkpoint as soon as the job is started, and then after the job is completed. Sometimes when the backlog is huge, the post-job checkpoint takes way too long and may not be executed at all. Since the time for which WAL files are retained is pretty low, WAL files may get deleted before the job that is processing them gets completed. In this case, there is data loss if the driver dies and restarts, because certain batches which were not completed at the time of checkpoint may have lost its WAL content as well.

This PR changes this to make sure that WAL cleanup happens only once the job is actually complete. This also changes the notion of what it means for a batch to be fully processed.

SparkQA commented:
[Test build #28439 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28439/consoleFull) for   PR 4964 at commit [`076b134`](https://github.com/apache/spark/commit/076b1344776f4e0d0242217aa0c0a823727994ab).
 * This patch merges cleanly.

harishreedharan commented:
I think a couple of tests will need to be reworked, which I will do if the logic looks ok.

SparkQA commented:
[Test build #28439 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28439/consoleFull) for   PR 4964 at commit [`076b134`](https://github.com/apache/spark/commit/076b1344776f4e0d0242217aa0c0a823727994ab).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28439/
Test FAILed.

harishreedharan commented:
Like I mentioned a couple of tests were failing. Once @tdas or @pwendell is OK with this idea, I will fix the tests.

harishreedharan committed:
[SPARK-6222]. Streaming jobs start a checkpoint as soon as the job is started, and then after
the job is completed. Sometimes when the backlog is huge, the post-job checkpoint takes way too
long and may not be executed at all. Since the time for which WAL files are retained is pretty
low, WAL files may get deleted before the job that is processing them gets completed. In this case,
there is data loss if the driver dies and restarts, because certain batches which were not
completed at the time of checkpoint may have lost its WAL content as well.

This PR changes this to make sure that WAL cleanup happens only once the job is actually complete.
This also changes the notion of what it means for a batch to be fully processed.

SparkQA commented:
[Test build #28457 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28457/consoleFull) for   PR 4964 at commit [`fa93b87`](https://github.com/apache/spark/commit/fa93b871ba0fe22924ff0273e975e492a6a7043c).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28457 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28457/consoleFull) for   PR 4964 at commit [`fa93b87`](https://github.com/apache/spark/commit/fa93b871ba0fe22924ff0273e975e492a6a7043c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28457/
Test PASSed.

squito commented:
I was just discussing this w/ Hari a bit, the one issue w/ this PR is having a test case to demonstrate the old problem, verify the fix, & prevent regression.  We have our own test cases which do that, but they aren't easy to port over here -- they involve forcefully killing the app to prevent some of the normal nice cleanup to happen.  Hari is checking to see if he can port them over here.  (I'm not clear on all the detals -- writing up the synopsis since he's about to hop on a plane.)

In any case, hopefully he'll find a better test case, or can elaborate on the difficulty.  But either way we're open to suggestions for testing.

harishreedharan committed:
To account for asynchronous execution, and not forgetting WAL files associated with a checkpoint, we need to ensure that the checkpoint completed after the oldest batch is the one we replay the WAL files from. So we need to keep track of the oldest batch which was completed and remove the WAL only once that batch's checkpoint is completed.

SparkQA commented:
[Test build #28522 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28522/consoleFull) for   PR 4964 at commit [`4d116e8`](https://github.com/apache/spark/commit/4d116e819dd39d4845b14ef4aea9de15604d526c).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28522 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28522/consoleFull) for   PR 4964 at commit [`4d116e8`](https://github.com/apache/spark/commit/4d116e819dd39d4845b14ef4aea9de15604d526c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28522/
Test PASSed.

tdas committed:
Fixing SPARK-6222

harishreedharan committed:
Update lastCompletedBatch only if the completed batch is newer than the previous one. If `concurrentJobs` is set to be > 1, then it is possible that the ordering of completion and checkpointing is different from the order in which they were started. So we need to be defensive against such a case.

SparkQA commented:
[Test build #28555 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28555/consoleFull) for   PR 4964 at commit [`a26c4e6`](https://github.com/apache/spark/commit/a26c4e6363ca69fec1ffd5154975459577fa4fab).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28555 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28555/consoleFull) for   PR 4964 at commit [`a26c4e6`](https://github.com/apache/spark/commit/a26c4e6363ca69fec1ffd5154975459577fa4fab).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28555/
Test FAILed.

harishreedharan committed:
Make lastCompletedBatch an Option.

SparkQA commented:
[Test build #28571 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28571/consoleFull) for   PR 4964 at commit [`b9d4bbc`](https://github.com/apache/spark/commit/b9d4bbc0b28fc9beda2fe3d2ba5eebd169246ae9).
 * This patch merges cleanly.

harishreedharan commented:
Looks like the tests passed, but the build has been stuck in MiMa for a while with lock not being available on the ivy cache:
`Waiting for lock on /home/jenkins/.ivy2/.sbt.ivy.lock to be available...`

SparkQA commented:
**[Test build #28571 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28571/consoleFull)**     for PR 4964 at commit [`b9d4bbc`](https://github.com/apache/spark/commit/b9d4bbc0b28fc9beda2fe3d2ba5eebd169246ae9)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28571/
Test FAILed.

shaneknapp commented:
jenkins, test this please

SparkQA commented:
[Test build #28581 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28581/consoleFull) for   PR 4964 at commit [`b9d4bbc`](https://github.com/apache/spark/commit/b9d4bbc0b28fc9beda2fe3d2ba5eebd169246ae9).
 * This patch merges cleanly.

harishreedharan commented:
Jenkins, please test this

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28581/
Test FAILed.

shaneknapp commented:
jenkins, test this please

SparkQA commented:
[Test build #28587 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28587/consoleFull) for   PR 4964 at commit [`b9d4bbc`](https://github.com/apache/spark/commit/b9d4bbc0b28fc9beda2fe3d2ba5eebd169246ae9).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28587 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28587/consoleFull) for   PR 4964 at commit [`b9d4bbc`](https://github.com/apache/spark/commit/b9d4bbc0b28fc9beda2fe3d2ba5eebd169246ae9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28587/
Test PASSed.

harishreedharan commented:
@tdas Please take a look.

tdas commented:
Will do. Sorry didnt get a chance today.
On Mar 13, 2015 11:54 PM, "Hari Shreedharan" <notifications@github.com>
wrote:

> @tdas <https://github.com/tdas> Please take a look.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/4964#issuecomment-80025583>.
>

harishreedharan commented:
When `concurrentJobs` is 1, we can add optimizations, but since both maps would be of size 1, I don't think the cost would be too high, but I am open to optimizing for that case.

SparkQA commented:
[Test build #28618 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28618/consoleFull) for   PR 4964 at commit [`67a5a50`](https://github.com/apache/spark/commit/67a5a500fe94e3d806a55ec19b88db2738f5e0c7).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28618 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28618/consoleFull) for   PR 4964 at commit [`67a5a50`](https://github.com/apache/spark/commit/67a5a500fe94e3d806a55ec19b88db2738f5e0c7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28618/
Test FAILed.

harishreedharan committed:
Further thread-safety improvements.

SparkQA commented:
[Test build #28622 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28622/consoleFull) for   PR 4964 at commit [`911379c`](https://github.com/apache/spark/commit/911379ce93464eb0371725cb0b156c0107f132d6).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28622 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28622/consoleFull) for   PR 4964 at commit [`911379c`](https://github.com/apache/spark/commit/911379ce93464eb0371725cb0b156c0107f132d6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28622/
Test PASSed.

harishreedharan commented:
@tdas - What do you think about adding a short-circuit for the scenario when `concurrentJobs == 1`? That would basically combine the current state of this PR with the simple approach in an earlier version of this PR: https://github.com/harishreedharan/spark/commit/fa93b871ba0fe22924ff0273e975e492a6a7043c

This would avoid access the `synchronized` blocks when only one job can be executed at a time.

tdas commented:
How about a combination of the https://github.com/harishreedharan/spark/commit/fa93b871ba0fe22924ff0273e975e492a6a7043c and my PR https://github.com/apache/spark/pull/5008?

harishreedharan commented:
But how would that combination really handle the case where `concurrentJobs > 1`? From what I can see we need either of https://github.com/harishreedharan/spark/commit/fa93b871ba0fe22924ff0273e975e492a6a7043c and #5008, since both solve this in pretty much the same way. 

I know that `concurrentJobs > 1` is not common, but it is non-zero. I have seen this being used in production.

tdas commented:
I thought your approach in
https://github.com/harishreedharan/spark/commit/fa93b871ba0fe22924ff0273e975e492a6a7043c
was to clear metadata on only the latest computed branch, rather than the
computed batch. We could do that in combination with the flag in my PR that
disables cleaning for the pre-batch-start checkpoint.



On Mon, Mar 16, 2015 at 1:01 PM, Hari Shreedharan <notifications@github.com>
wrote:

> But how would that combination really handle the case where concurrentJobs
> > 1? From what I can see we need either of harishreedharan@fa93b87
> <https://github.com/harishreedharan/spark/commit/fa93b871ba0fe22924ff0273e975e492a6a7043c>
> and #5008 <https://github.com/apache/spark/pull/5008>, since both solve
> this in pretty much the same way.
>
> I know that concurrentJobs > 1 is not common, but it is non-zero. I have
> seen this being used in production.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/4964#issuecomment-81905729>.
>

harishreedharan commented:
So in that version of the PR, I remove the metadata in the checkpoint only after the batch has been completed. So what it does is:
- when a batch is completed, update `lastCompletedBatch`
- when a checkpoint is completed, if the time is same as `lastCompletedBatch`, then clear metadata, else don't - so pre-batch-start checkpoint does not clean the metadata anyway. (I can admittedly use better way of comparing, rather than the `Seq.min`

This gets the same result as #5008, but it is not enough when we could have multiple jobs running at once. When multiple jobs are running at once, we need to keep track of which jobs have been started and clear the metadata only for the oldest batch that has been completed and their post-completion checkpoint are complete. Any batch newer than that one, which may have completed and then checkpointed should not clean any metadata - that is what this PR's current state is. (There is some code to move the oldest processed batch forward once the oldest one has been processed - the iterator stuff, so any newer batches which have been completed and checkpoint can be removed from tracking.)

harishreedharan committed:
Switch checkpointsCompleted to a HashMap.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -156,14 +164,21 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
    * Callback called when a batch has been completely processed.
    */
-  def onBatchCompletion(time: Time) {
+  def onBatchCompletion(time: Time): Unit = synchronized {
```
Why the synchronized at all?

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/Time.scala:
```diff
@@ -81,6 +81,18 @@ case class Time(private val millis: Long) {
 
   override def toString: String = (millis.toString + " ms")
 
+  override def hashCode: Int = {
```
Dude, Time is a case class, this is not necessary :)

SparkQA commented:
[Test build #28671 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28671/consoleFull) for   PR 4964 at commit [`f41d23c`](https://github.com/apache/spark/commit/f41d23c4d9d86c8d67498481d866683d741f0450).
 * This patch merges cleanly.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -43,6 +47,9 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
   private val ssc = jobScheduler.ssc
   private val conf = ssc.conf
   private val graph = ssc.graph
+  private lazy val completedBothCheckpoints = new HashMap[Time, Boolean]()
```
I dont like the fact that the data structure encodes that there will be two checkpoints for each batch. Checkpointing is something that should be doable any time, and any number of times. Restricting a data strcuture to a constraint that it can only be done twice is bad design.

In fact I say that my earlier design of automatically deleting checkpoints whenever checkpoint is called, was wrong as well. That's why adding the flag in my PR is important, so that we dissociate the act of taking a checkpoint from the act of clearing checkpoints.

harishreedharan replied:
Ah, makes sense. Do you want me to pull in your commit into this one? That makes sense.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -43,6 +47,9 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
   private val graph = ssc.graph
+  private lazy val completedBothCheckpoints = new HashMap[Time, Boolean]()
+  private lazy val batchesGenerated = new TreeSet[Time](new TimeComparator)
+  private val noConcurrentJobs = ssc.sc.conf.getInt("spark.streaming.concurrentJobs", 1) == 1
```
I dont like this, but it could be a necessary evil to make sure most users are unaffected by any change here.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -254,39 +271,97 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
 
     // If checkpointing is enabled, then checkpoint,
     // else mark batch to be fully processed
-    if (shouldCheckpoint) {
```
Why this line change. `shouldCheckpoint` can stay as is.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -254,39 +271,97 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
+    }
+  }
+
+  private class TimeComparator extends Comparator[Time] {
```
Better to put this in the same file as Time.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -156,14 +165,21 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
    * Callback called when the checkpoint of a batch has been written.
    */
-  def onCheckpointCompletion(time: Time) {
+  def onCheckpointCompletion(time: Time): Unit = synchronized {
```
Again no need of synchronized.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -17,9 +17,13 @@
 
-import akka.actor.{ActorRef, Props, Actor}
+import akka.actor.{Actor, ActorRef, Props}
+import com.google.common.primitives.Longs
```
What is this for?

SparkQA commented:
[Test build #28671 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28671/consoleFull) for   PR 4964 at commit [`f41d23c`](https://github.com/apache/spark/commit/f41d23c4d9d86c8d67498481d866683d741f0450).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28671/
Test PASSed.

harishreedharan committed:
Removing a whole lot synchronization, clarifying names of variables etc.

SparkQA commented:
[Test build #28687 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28687/consoleFull) for   PR 4964 at commit [`8acb827`](https://github.com/apache/spark/commit/8acb827706dfada00cd2e58a128d2abacb65d904).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28687 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28687/consoleFull) for   PR 4964 at commit [`8acb827`](https://github.com/apache/spark/commit/8acb827706dfada00cd2e58a128d2abacb65d904).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class CheckpointWriteHandler(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28687/
Test FAILed.

harishreedharan committed:
Cleanup even checkpoint and metadata defensively

SparkQA commented:
[Test build #28688 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28688/consoleFull) for   PR 4964 at commit [`1022ccf`](https://github.com/apache/spark/commit/1022ccf40557961b00fbc7c88a559909d7206a09).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28688 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28688/consoleFull) for   PR 4964 at commit [`1022ccf`](https://github.com/apache/spark/commit/1022ccf40557961b00fbc7c88a559909d7206a09).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class CheckpointWriteHandler(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28688/
Test FAILed.

harishreedharan committed:
Fixing style issues.

SparkQA commented:
[Test build #28693 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28693/consoleFull) for   PR 4964 at commit [`030efd7`](https://github.com/apache/spark/commit/030efd72d9a7e071b00d0519397b497c3082d598).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28693 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28693/consoleFull) for   PR 4964 at commit [`030efd7`](https://github.com/apache/spark/commit/030efd72d9a7e071b00d0519397b497c3082d598).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class CheckpointWriteHandler(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28693/
Test FAILed.

harishreedharan committed:
Moving clearMetadata back to the original place. It can be regenerated later anyway.

SparkQA commented:
[Test build #28699 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28699/consoleFull) for   PR 4964 at commit [`b4f807b`](https://github.com/apache/spark/commit/b4f807bd6c7fff43b1b2bb72a8d53de82ae24da1).
 * This patch merges cleanly.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -44,6 +49,14 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
+  // Scala treeset makes it difficult to drop items while iterating over the set. dropWhile
+  // creates a new set, which does not seem very efficient. So use Java treeset instead.
+  private val activeBatches =
+    Collections.synchronizedSet[Time](new TreeSet[Time](new TimeComparator()))
```
What happened to using Scala Treeset?

harishreedharan replied:
Scala treeset does not have a way of dropping data based on some predicate. `dropWhile` seems inefficient since it creates a new set - which seemed unnecessary. An alternative would be using the iterator to drop, but the scala iterator does not allow removing data.

SparkQA commented:
[Test build #28699 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28699/consoleFull) for   PR 4964 at commit [`b4f807b`](https://github.com/apache/spark/commit/b4f807bd6c7fff43b1b2bb72a8d53de82ae24da1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class CheckpointWriteHandler(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28699/
Test PASSed.

tdas commented:
Offline discussion with @harishreedharan : There are cases that this fix does not cover, and doing a principled approach that handles the case where concurrentJobs > 1 (which is not officially supported) requires a non-trivial refactoring. In fact, for concurrentJobs > 1, there were existing bugs independent of the regression referred to in SPARK-6222. So that makes this issue not really a regression for officially non-supported cases of concurrentJobs > 1. To fix the regression SPARK-6222 for the the default, officially supported case of concurrentJobs = 1, we are going to use the surgical fix in PR #5008.

tdas commented:
This PR is not necessary any more in its current form. Mind closing this?
